compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt

```diff
@@ -386,23 +392,64 @@ private val jvmFilePhases = listOf(
     // makePatchParentsPhase()
 )
 
-val jvmLoweringPhases = NamedCompilerPhase(
-    name = "IrLowering",
-    description = "IR lowering",
-    nlevels = 1,
-    actions = setOf(defaultDumper, validationAction),
-    lower = validateIrBeforeLowering then
-            processOptionalAnnotationsPhase then
-            expectDeclarationsRemovingPhase then
-            serializeIrPhase then
-            scriptsToClassesPhase then
-            fileClassPhase then
-            jvmStaticInObjectPhase then
-            repeatedAnnotationPhase then
-            performByIrFile(lower = jvmFilePhases) then
-            generateMultifileFacadesPhase then
-            resolveInlineCallsPhase then
-            // should be last transformation
-            prepareForBytecodeInlining then
-            validateIrAfterLowering
-)
+val jvmLoweringPhases = buildJvmLoweringPhases("IrLowering", listOf("PerformByIrFile" to jvmFilePhases))
+
+private fun buildJvmLoweringPhases(
+    name: String,
+    phases: List<Pair<String, List<NamedCompilerPhase<JvmBackendContext, IrFile>>>>
+): NamedCompilerPhase<JvmBackendContext, IrModuleFragment> {
+    return NamedCompilerPhase(
+        name = name,
+        description = "IR lowering",
+        nlevels = 1,
+        actions = setOf(defaultDumper, validationAction),
+        lower =
+        fragmentSharedVariablesLowering then
+                validateIrBeforeLowering then
+                processOptionalAnnotationsPhase then
+                expectDeclarationsRemovingPhase then
+                serializeIrPhase then
+                scriptsToClassesPhase then
+                fileClassPhase then
+                jvmStaticInObjectPhase then
+                repeatedAnnotationPhase then
+                buildLoweringsPhase(phases) then
+                generateMultifileFacadesPhase then
+                resolveInlineCallsPhase then
+                // should be last transformation
+                prepareForBytecodeInlining then
+                validateIrAfterLowering
+    )
+}
+
+private fun buildLoweringsPhase(
```

Comment:
Sorry, I couldn't resist rewriting it a bit more in a functional style. It's not perfect though because requires explicit generic type arguments. :/
```
private fun buildLoweringsPhase(
    perModuleLowerings: List<Pair<String, List<NamedCompilerPhase<JvmBackendContext, IrFile>>>>,
): CompilerPhase<JvmBackendContext, IrModuleFragment, IrModuleFragment> =
    perModuleLowerings.map { (name, lowerings) -> performByIrFile(name, lower = lowerings) }
        .reduce<
                CompilerPhase<JvmBackendContext, IrModuleFragment, IrModuleFragment>,
                CompilerPhase<JvmBackendContext, IrModuleFragment, IrModuleFragment>
                > { result, phase -> result then phase }
```

Reply:
I can always appreciate a good fold! Done!